include/hipSYCL/sycl/libkernel/detail/vec.hpp

```diff
@@ -78,20 +78,82 @@ template<> struct logical_vector_op_result<detail::dp_float>
 
 template <int...> struct vector_index_sequence {};
 
+namespace emulated_types {
+
+using uchar = unsigned char;
+using ushort = unsigned short;
+using uint = unsigned int;
+using longlong = long long;
+using ulong = unsigned long;
+using ulonglong = unsigned long long;
+
+template <class T, int Dim> struct required_alignment {
+#if defined(HIPSYCL_PLATFORM_CUDA)
+  constexpr static size_t value = alignof(T) * Dim;
+#else
+  // Use default alignment otherwise
+  constexpr static size_t value = 0;
+#endif
+};
+
+// TODO: This should probably be moved to a backend-specific location.
+#if defined(HIPSYCL_PLATFORM_CUDA)
+template <class T> struct required_alignment<T, 3>
+{ constexpr static size_t value = alignof(T); };
+template <> struct required_alignment<long, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<ulong, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<longlong, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<ulonglong, 4>
+{ constexpr static size_t value = 16; };
+template <> struct required_alignment<double, 4>
+{ constexpr static size_t value = 16; };
```

Comment:
Why do we need to hardcode the alignments? Couldn't we just set them to `alignof(NativeT)`, with `NativeT` being native CUDA/HIP vector types? Then it would automatically do the right thing on ROCm as  well.

Reply:
Uhh - to be honest I didn't really think too much about how the hipSYCL compilation model works for the new runtime. From the way the vector implementation was previously structured (with branching on `SYCL_DEVICE_ONLY`), I assumed that there was some phase of compilation where the native types wouldn't be available -- hence I also couldn't do something like `alignof(NativeT)`. But thinking about it now, that probably doesn't make any sense for CUDA/HIP, as they also support the native vector types in host code.

I realize the `emulated_vector` types are required for backends that don't have native vector types (i.e., OpenMP), but now my question is: Why use the emulated types at all when targeting CUDA/HIP?

Reply:
Great point - it seems we missed the best solution :) You are right, we now need to distinguish two cases:
1) CPU backend embedded in host pass of CUDA/HIP
2) Standalone CPU backend

For 2), the alignment won't matter (but it might make sense to align for vectorization), and we don't need to match alignments on other backends. For 1) we should be able to use the device types also on host, which should also solve the alignment issue.

Reply:
Okay, so it should be sufficient to check for `#if defined(HIPSYCL_PLATFORM_ROCM) || defined(HIPSYCL_PLATFORM_CUDA)` for the first case, and the `#else` branch would be the second one, right?

By the way, it might still be worth considering having consistent alignment requirements across all backends. Otherwise dumping a buffer to file and reading it back in (on another backend) might lead to surprises. Does the 2020 spec say anything about this? Anyway, I guess that's out of scope for this PR.

Reply:
> Okay, so it should be sufficient to check for #if defined(HIPSYCL_PLATFORM_ROCM) || defined(HIPSYCL_PLATFORM_CUDA) for the first case, and the #else branch would be the second one, right?

Yes. It might be a good idea to add a macro to test this more easily to `libkernel/backend.hpp` so that we only need to change that one place in case we add more backends in the future.

> By the way, it might still be worth considering having consistent alignment requirements across all backends. Otherwise dumping a buffer to file and reading it back in (on another backend) might lead to surprises. Does the 2020 spec say anything about this?

Good question. I checked and it seems SYCL 2020 provisional specifies `vec` alignments explicitly in 4.16.2.6. Which would mean that we cannot just rely on CUDA/HIP types to do the right thing. For a fully compliant implementation, we may have to use our own types for all backends.